File: js/test/htdocs/spec/test_conversions.js

describe('Conversions', function() {

    it('Pack/unpack bytes: Identity', function() {
		const test = new TestConversions();
		
		// Check some examples
        test.testPackUnpack([]);

        test.testPackUnpack([0]);
        test.testPackUnpack([1]);
        test.testPackUnpack([10]);
        test.testPackUnpack([255]);

        test.testPackUnpack([255, 0]);
        test.testPackUnpack([0, 255]);

        test.testPackUnpack([0, 127, 255]);
        test.testPackUnpack([255, 255, 255]);
        test.testPackUnpack([0, 0, 0]);
        test.testPackUnpack([127, 127, 127]);

        test.testPackUnpack([2, 0, 3, 4, 255, 200]);
        test.testPackUnpack([2, 0, 3, 4, 0]);
        test.testPackUnpack([2, 0, 3, 4, 255, 200, 222]);        
    });
    
    it('Pack/unpack bytes: Sizes', function() {
		// Check size of packed content
        const bridge = new JsMidiBridge();

        expect(bridge.packBytes(new Uint8Array([255])).length).toBe(2);
        expect(bridge.packBytes(new Uint8Array([255, 255])).length).toBe(3);
        expect(bridge.packBytes(new Uint8Array([255, 255, 255])).length).toBe(4);
        expect(bridge.packBytes(new Uint8Array([255, 255, 255, 255])).length).toBe(5);
        expect(bridge.packBytes(new Uint8Array([255, 255, 255, 255, 255])).length).toBe(6);
        expect(bridge.packBytes(new Uint8Array([255, 255, 255, 255, 255, 255])).length).toBe(7);
        expect(bridge.packBytes(new Uint8Array([255, 255, 255, 255, 255, 255, 255])).length).toBe(8);
        expect(bridge.packBytes(new Uint8Array([255, 255, 255, 255, 255, 255, 255, 255])).length).toBe(10);
	});
	
	
	//##########################################################################################################


	it('String conversion: Identity', function() {
		const test = new TestConversions();
		
		test.testStringConversion("");
        test.testStringConversion(" ");
        test.testStringConversion("      ");

        test.testStringConversion("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"§$%&/()=?*+#-_.:,;<>^°`´");
        test.testStringConversion("äöüÖÄÜß");
	});
	
	
	//##########################################################################################################
	
	
	it('Number conversion: Identity', function() {
		const test = new TestConversions();
		
		for (let i=0; i<99; ++i) {
            test.testNumberConversion(i, 1, 2);
        }

        for (let i=0; i<256; ++i) {
            test.testNumberConversion(i, 2, 3);
		}
		
        for (let i=0; i<9999; i += 10) {
            test.testNumberConversion(i, 3, 4);
            test.testNumberConversion(i, 4, 5);
            
            // Note: JS does have problems shifting outside the 32bit border, so we do not support this here
            // (which is not problem as the bridge only needs 3 byte integers, but could be upped to 4 in the future) 
        }
	});
});


//##########################################################################################################


class TestConversions {
	
	testPackUnpack(data) {
	    const bridge = new JsMidiBridge();
	    
	    data = new Uint8Array(data);
	            
	    const packed = bridge.packBytes(data);
	    
	    this.checkHalfBytes(packed);
	
	    if (data.size > 0) {
	        expect(data).not.toEqual(packed);
	    }
	
	    const unpacked = bridge.unpackBytes(packed);
	
		expect(unpacked).toEqual(data);
	}
	
	checkHalfBytes(data) {
	    expect(data).toBeInstanceOf(Uint8Array);
	    
		for (const b of data) {
			expect(b).toBeGreaterThanOrEqual(0);
			expect(b).toBeLessThan(128);
	    }
	}
	
	testStringConversion(str) {
	    const bridge = new JsMidiBridge();
	
	    const encoded = bridge.string2bytes(str);
	    
	    this.checkHalfBytes(encoded);
	
	    const decoded = bridge.bytes2string(encoded);
	
	    expect(decoded).toEqual(str);
	}
	
	testNumberConversion(value, numBytes, outputLen) {
	    const bridge = new JsMidiBridge();
	
	    const encoded = bridge.number2bytes(value, numBytes);
	
		expect(encoded.length).toBe(outputLen);
	    
	    this.checkHalfBytes(encoded);
	    
	    const decoded = bridge.bytes2number(encoded);
	
		expect(decoded).toBe(value);
	}	
}

File: python/test/pymidibridge/test_conversions.py
import unittest

# Import subject under test
from lib.pymidibridge.pymidibridge import PyMidiBridge

from .mocks import *


class TestConversions(unittest.TestCase):

    def test_pack_unpack_bytes(self):
        # Check some examples
        self._test_pack_unpack_bytes([])

        self._test_pack_unpack_bytes([0])
        self._test_pack_unpack_bytes([1])
        self._test_pack_unpack_bytes([10])
        self._test_pack_unpack_bytes([255])

        self._test_pack_unpack_bytes([255, 0])
        self._test_pack_unpack_bytes([0, 255])

        self._test_pack_unpack_bytes([0, 127, 255])
        self._test_pack_unpack_bytes([255, 255, 255])
        self._test_pack_unpack_bytes([0, 0, 0])
        self._test_pack_unpack_bytes([127, 127, 127])

        self._test_pack_unpack_bytes([2, 0, 3, 4, 255, 200])
        self._test_pack_unpack_bytes([2, 0, 3, 4, 0])
        self._test_pack_unpack_bytes([2, 0, 3, 4, 255, 200, 222])


    def _test_pack_unpack_bytes(self, data):
        bridge = PyMidiBridge(None, None)
                
        packed = bridge._pack_bytes(bytes(data))
        
        self._check_half_bytes(packed)

        if data:
            self.assertNotEqual(data, packed)

        unpacked = bridge._unpack_bytes(packed)

        self.assertEqual(unpacked, bytes(data))


    def _check_half_bytes(self, data):
        for b in list(data):
            self.assertGreaterEqual(b, 0)
            self.assertLess(b, 128)


    def test_pack_bytes_sizes(self):
        # Check size of packed content
        bridge = PyMidiBridge(None, None)

        self.assertEqual(len(bridge._pack_bytes(bytes([255]))), 2)           
        self.assertEqual(len(bridge._pack_bytes(bytes([255, 255]))), 3)      
        self.assertEqual(len(bridge._pack_bytes(bytes([255, 255, 255]))), 4) 
        self.assertEqual(len(bridge._pack_bytes(bytes([255, 255, 255, 255]))), 5)
        self.assertEqual(len(bridge._pack_bytes(bytes([255, 255, 255, 255, 255]))), 6)
        self.assertEqual(len(bridge._pack_bytes(bytes([255, 255, 255, 255, 255, 255]))), 7)
        self.assertEqual(len(bridge._pack_bytes(bytes([255, 255, 255, 255, 255, 255, 255]))), 8)
        self.assertEqual(len(bridge._pack_bytes(bytes([255, 255, 255, 255, 255, 255, 255, 255]))), 10)


    ##########################################################################################################


    def test_string_conversion(self):
        self._test_string_conversion("")
        self._test_string_conversion(" ")
        self._test_string_conversion("      ")

        self._test_string_conversion("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"§$%&/()=?*+#-_.:,;<>^°`´")
        self._test_string_conversion("äöüÖÄÜß")


    def _test_string_conversion(self, str):
        bridge = PyMidiBridge(None, None)

        encoded = bridge._string_2_bytes(str)    

        self._check_half_bytes(encoded)

        decoded = bridge._bytes_2_string(encoded)

        self.assertEqual(decoded, str)


    ##########################################################################################################


    def test_number_conversion(self):
        for i in range(99):
            self._test_number_conversion(i, 1, 2)

        for i in range(256):
            self._test_number_conversion(i, 2, 3)

        for i in range(0, 9999, 10):
            self._test_number_conversion(i, 3, 4)
            self._test_number_conversion(i, 5, 6)
            self._test_number_conversion(i, 7, 8)


    def _test_number_conversion(self, value, num_bytes, output_len):
        bridge = PyMidiBridge(None, None)

        encoded = bridge._number_2_bytes(value, num_bytes)

        self.assertEqual(len(encoded), output_len)
        self._check_half_bytes(encoded)

        decoded = bridge._bytes_2_number(encoded)

        self.assertEqual(decoded, value, "Value: " + repr(value) + "; Num Bytes: " + repr(num_bytes) + " Packed: " + repr(encoded))

